src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -3776,6 +3780,99 @@ public byte byteValueExact() {
        return (byte)num;
     }
 
+    /**
+     * Converts this {@code BigDecimal} to a {@code Float16}.
+     * This conversion is similar to the
+     * <i>narrowing primitive conversion</i> from {@code double} to
+     * {@code float} as defined in
+     * <cite>The Java Language Specification</cite>:
+     * if this {@code BigDecimal} has too great a
+     * magnitude to represent as a {@code Float16}, it will be
+     * converted to {@link Float16#NEGATIVE_INFINITY} or {@link
+     * Float16#POSITIVE_INFINITY} as appropriate.  Note that even when
+     * the return value is finite, this conversion can lose
+     * information about the precision of the {@code BigDecimal}
+     * value.Float16
+     *
+     * @return this {@code BigDecimal} converted to a {@code Float16}.
+     * @jls 5.1.3 Narrowing Primitive Conversion
+     */
+    public Float16 float16Value() {
+        /* For details, see the extensive comments in doubleValue(). */
+        if (intCompact != INFLATED) {
+            Float16 v = Float16.valueOf(intCompact);
+            if (scale == 0) {
+                return v;
+            }
+            /*
+             * The discussion for the double case also applies here. That is,
+             * the following test is precise for all long values, but here
+             * Long.MAX_VALUE is not an issue.
+             */
+            if (v.longValue() == intCompact) {
+                if (0 < scale && scale < FLOAT16_10_POW.length) {
+                    return Float16.divide(v, FLOAT16_10_POW[scale]);
+                }
+                if (0 > scale && scale > -FLOAT16_10_POW.length) {
+                    return Float16.multiply(v, FLOAT16_10_POW[-scale]);
+                }
+            }
+        }
+        return fullFloat16Value();
+    }
+
+    private Float16 fullFloat16Value() {
+        if (intCompact == 0) {
+            return Float16.valueOf(0);
+        }
+        BigInteger w = unscaledValue().abs();
+        long qb = w.bitLength() - (long) Math.ceil(scale * L);
+        if (qb < Q_MIN_F16 - 2) {  // qb < -26
+            return Float16.valueOf(signum() * 0.0f);
+        }
+        if (qb > Q_MAX_F16 + P_F16 + 1) {  // qb > 17
+            return Float16.valueOf(signum() * Float.POSITIVE_INFINITY);
+        }
+        if (scale < 0) {
+            return Float16.valueOf(signum() * w.multiply(bigTenToThe(-scale)).floatValue());
```

Comment:
Are the calls to `floatValue()` for the `scale < 0` and `scale == 0` cases meant to be calls to `float16Value()`?

If not, the different coding pattern from the analagous float and double methods should be documented in a comment.

Reply:
Improved code to mirror analogous code in `fullDoubleValue()`.